libraries/chain/controller.cpp

```diff
@@ -4356,15 +4358,24 @@ struct controller_impl {
             }
          }
 
+         auto start = fc::time_point::now();
          for( auto ritr = new_head_branch.rbegin(); ritr != new_head_branch.rend(); ++ritr ) {
             auto except = std::exception_ptr{};
             const auto& bsp = *ritr;
             try {
                bool applied = apply_block( bsp, bsp->is_valid() ? controller::block_status::validated
                                                                 : controller::block_status::complete, trx_lookup );
-               if (!switch_fork && (!applied || check_shutdown())) {
-                  shutdown();
-                  break;
+               if (!switch_fork) { // always complete a switch fork
+                  if (!applied || check_shutdown()) {
+                     shutdown();
+                     break; // result should be complete since we are shutting down
+                  }
+                  // Break every ~500ms to allow other tasks (e.g. get_info, SHiP) opportunity to run. User expected
+                  // to call apply_blocks again if this returns incomplete.
+                  if (!replaying && fc::time_point::now() - start > fc::milliseconds(500)) {
```

Comment:
do we need this `!replaying` check since the callers of `maybe_apply_blocks` now call it in a while loop?

Reply:
If replaying there is no reason to come up for air. Nothing else can run.

Reply:
So can it be interrupted wit Ctrl-C?

Reply:
Yes. I verfied that it works.

Reply:
How does it happen? Just curious I don't see it.

Reply:
> How does it happen? Just curious I don't see it. I see the ` check_shutdown()` at line 4369, but won't we have the issue that the main thread being busy doesn't process the signal?

Oh I get it, it is because the replay is called by `startup()`, which is called by `chain_plugin_impl::plugin_startup()` where we still have the separate thread catching signals.